bin/proving-service/src/proxy/worker.rs

```diff
@@ -1,75 +1,216 @@
+    },
 }
 
 impl Worker {
-    /// Creates a new worker and a gRPC health check client for the given worker address.
+    /// Creates a new worker and a gRPC status client for the given worker address.
     ///
     /// # Errors
     /// - Returns [ProvingServiceError::InvalidURI] if the worker address is invalid.
     /// - Returns [ProvingServiceError::ConnectionFailed] if the connection to the worker fails.
     pub async fn new(
-        worker: Backend,
+        worker_addr: String,
         connection_timeout: Duration,
         total_timeout: Duration,
     ) -> Result<Self, ProvingServiceError> {
-        let health_check_client =
-            create_health_check_client(worker.addr.to_string(), connection_timeout, total_timeout)
-                .await?;
+        let backend =
+            Backend::new(&worker_addr).map_err(ProvingServiceError::BackendCreationFailed)?;
+
+        let status_client =
+            create_status_client(worker_addr, connection_timeout, total_timeout).await?;
 
         Ok(Self {
-            backend: worker,
+            backend,
             is_available: true,
-            health_check_client,
+            status_client,
+            health_status: WorkerHealthStatus::Healthy,
+            version: "".to_string(),
         })
     }
 
+    /// Returns the worker address.
     pub fn address(&self) -> String {
         self.backend.addr.to_string()
     }
 
-    pub async fn is_healthy(&mut self) -> bool {
-        match self
-            .health_check_client
-            .check(HealthCheckRequest { service: "".to_string() })
-            .await
-        {
-            Ok(response) => response.into_inner().status() == ServingStatus::Serving,
-            Err(err) => {
-                error!("Failed to check worker health ({}): {}", self.address(), err);
-                false
+    /// Checks the worker status.
+    ///
+    /// # Returns
+    /// - `Some(true)` if the worker is ready.
+    /// - `Some(false)` if the worker is not ready or if there was an error checking the status.
+    /// - `None` if the worker should not do a health check.
+    pub async fn check_status(
+        &mut self,
+        supported_prover_type: &ProverType,
+    ) -> Option<WorkerHealthStatus> {
+        if !self.should_do_health_check() {
+            return None;
+        }
+
+        let failed_attempts = self.retries_amount();
+
+        let worker_status = match self.status_client.status(StatusRequest {}).await {
+            Ok(response) => response.into_inner(),
+            Err(e) => {
+                error!("Failed to check worker status ({}): {}", self.address(), e);
+                return Some(WorkerHealthStatus::Unhealthy {
+                    failed_attempts: failed_attempts + 1,
+                    first_fail_timestamp: Instant::now(),
+                    reason: e.message().to_string(),
+                });
             },
+        };
+
+        self.version = worker_status.version;
```

Comment:
We should also check version compatibility. Basically, if the proxy is at v0.8.0 and worker is at v0.9.0, the worker should be marked as "unhealthy".